0000-experiment-dataframe.md

```diff
@@ -0,0 +1,320 @@
+
+`ExperimentData.add_analysis_results()` dispatches the add logic.
+The method signature is updated to `results: Optional[Union[AnalysisResult, List[AnalysisResult]]] = None, **kwargs: Any`.
+If results argument is not None, this must be the legacy analysis result data.
+This implies existing analysis classes can still return this legacy object.
+Legacy result objects are internally converted into a single row of the data frame and added.
+If there are only keyword arguments, we can directly populate the analysis result data frame and bypasses generation of redundant `AnalysisResult` which no longer exists in the `ExperimentData`.
+Existing analysis classes in the Qiskit Experiments namespace will switch to this kwargs style, namely, it must directly populate analysis data from `_run_analysis` method.
+
+`ExperimentData.artifacts()` and `.add_artifacts()` methods are newly implemented with new protected member `ExpeirmentData._artifacts` which is a thread safe list of artifacts.
+This method may provide a filtering function based on the artifact metadata.
+
+_Deprecations_
+- `CurveData` will be replaced with the data frame and thus immediately deprecated. This is in internal use for curve analysis, thus the impact is minimum.
+- Curve analysis options `return_data_points` (curve data) and `return_fit_parameters` (fit status). Now these are moved to artifacts rather than in the analysis results.
+- Add deprecation warning when `ExpeirmentData.analysis_results()` is called with `dataframe=False` (default).
+
+_Remove_
+- None
+
+##### qiskit_experiments v0.7
+In this version we switch the default option of `ExpeirmentData.analysis_results()` to `dataframe=None`. It returns data frame by defaults.
+We must write analysis classes so that it adds analysis results with keyword arguments.
+
+_Deprecations_
+- `AnalysisResult` and `AnalysisResultData`.
+- Deprecate calling `ExperimentData.add_analysis_results()` with results argument.
+- Calling with `ExpeirmentData.analysis_results()` with dataframe argument.
+
+_Remove_
+- `CurveData`
+- Remove fit status and curve data entry generation in analysis results. Curve analysis will complete migration to the artifact at this point.
+
+##### qiskit_experiments v0.8
+In this version we complete migration to data frame and legacy QE classes and methods are completely removed.
+
+_Deprecations_
+- None
+
+_Remove_
+- `AnalysisResult` and `AnalysisResultData`.
+- results argument from the `ExpeirmentData.add_analysis_results()`.
+- dataframe argument from the `ExpeirmentData.analysis_results()`.
+
+
+## Alternative Approaches
+N/A
+
+## Questions
+- What data frame columns must be shown to end users.
+- How we expose `extra` column? A single column of a free-form dict is enough? We cannot apply convenient data frame filtering methods to such column. This must be flattened? This may cause problem when we have multiple experiments, i.e. BatchExperiment. Every experiment may have different extra dict keys.
```

Comment:
Maybe we could have defined columns for the service, and then a compacted column (ie extra), that can be expanded / collapsed when loading and saving? For batch experiments this might involve a lot of columns as the union of all individual results, where some would need to have a empty/null/ N/A field to distinguish which should be discarded when collapsing back down.

If we do something like this im not sure if this functionality should be in the service, or qiskit experiments. (It could probably be done with custom accessors too)

Reply:
> For batch experiments this might involve a lot of columns as the union of all individual results, where some would need to have a empty/null/ N/A field to distinguish which should be discarded when collapsing back down.

Yes, that's my concern. Probably, QE can always format data frame so that it only returns union of columns to end users, e.g.

```
entry1: ["value", "units", "chisq", "components", ...]
entry2: ["value", "units", "chisq", "components", ...]
entry3: ["value", "components", ...]
```

When a user request data of entry1 and entry2, QE (`ExperimentData`) returns `"value", "units", "chisq", "components",...`.  When one requests 1 and 3, only `"value", "components", ...` are returned.